✨ User 기본 기능 구현

tellang opened the pull request:
## 🔗 관련 이슈

Closes #16

## 📌 PR 요약

사용자 서비스 초기화 및 관련 유틸리티 추가

## 🔍 주요 변경 사항

- 사용자 서비스 초기화 및 테스트 설정
- 사용자 엔티티, 레포지토리 및 DTO 추가
- 공통 에러 및 성공 코드 추가
- 사용자 테스트 유틸리티 추가 (H2DbCleaner 등)

## 🧪 테스트 방법

```bash
./gradlew test
```

## ✅ PR 체크리스트

- [x] 코드 컨벤션을 준수했나요?
- [x] 필요한 테스트를 추가하고 모든 테스트가 통과하나요?
- [ ] 관련 문서를 업데이트했나요? (필요한 경우)

## 👀 리뷰어 체크 포인트

- User 엔티티 및 레포지토리의 구조 확인
- TestUserEntityFactory와 관련된 테스트 유틸리티 검토

## 💬 기타 코멘트
- 추후 작업
  - 빈 벨리데이션 추가
  - 컨트롤러 구현
    - 해당 부분 이슈는 분리 예정입니다
  - search 부분 구현
- slack 으로 공유될 user-local.env 파일을 환경 변수로 반드시 등록해야 합니다.
- 추후 모듈분리를 염두에 두고 작성되었습니다.

tellang committed:
:tada: 사용자 서비스 프로젝트 초기화 (#16)

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:white_check_mark: 사용자 서비스 프로젝트 테스트 및 설정 파일 추가 (#16)

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:sparkles: 사용자 엔티티 및 베이스 엔티티 추가 (#16)

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:sparkles: 공통 에러 및 성공 코드 추가 (#16)

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:sparkles: 사용자 레포지토리 및 헬퍼 클래스 추가 (#16)

- UserRepository.java를 추가하여 사용자 엔티티와 연동된 JPA 레포지토리를 정의합니다.
- RepositoryHelper 인터페이스와 AbstractRepositoryHelper 추상 클래스를 추가하여 기본적인 레포지토리 헬퍼 구조를 제공합니다.
- UserRepositoryHelper.java를 추가하여 사용자 엔티티의 데이터베이스 접근을 위한 헬퍼 클래스를 구현합니다.
- RoleType.java는 사용자 역할을 관리하며, USER, MANAGER, MASTER 등의 역할을 정의합니다.

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:white_check_mark: H2 데이터베이스 클리너 및 테스트 유틸 추가 (#16)

- H2DbCleaner.java를 추가하여 테스트 중에 H2 데이터베이스의 모든 테이블을 정리하는 유틸리티 클래스를 제공합니다. 이를 통해 테스트 실행 시 데이터베이스 상태를 매번 초기화합니다.
- ArbitraryField.java를 추가하여 테스트에서 사용할 상수 값(예: 사용자 이름, 전화번호, 이메일 등)을 정의합니다. 이를 통해 테스트에서 중복되는 필드 값을 쉽게 관리할 수 있습니다.

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:sparkles: 사용자 서비스 및 DTO 추가 (#16)

- UserService.java에 사용자 생성, 조회, 삭제, 소프트 삭제 등의 비즈니스 로직을 구현합니다.
- UserDto.java를 통해 사용자 데이터 전송 객체를 정의합니다. 이는 사용자 데이터를 외부로 노출할 때 사용되며, Create, Result, Delete.Soft 등의 구조를 포함합니다.
- UserMapper.java는 DTO와 엔티티 간의 변환을 담당하여, 비즈니스 로직에서 이를 활용합니다.

Signed-off-by: tellang <tellang@daum.net>

tellang committed:
:white_check_mark: 사용자 엔티티 생성 헬퍼 추가 (#16)

- `TestUserEntityFactory.java`를 추가하여 사용자 엔티티를 테스트 환경에서 생성하고 초기화할 수 있습니다.
- `UserServiceTests.java`는 `UserService`의 각 메서드가 올바르게 동작하는지 테스트하는 JUnit 테스트 클래스로, 사용자 생성, 조회, 삭제 관련 테스트를 포함합니다.
- `ArbitraryUserFactory.java`를 추가하여 테스트 환경에서 사용자 DTO와 엔티티를 쉽게 생성할 수 있습니다.

Signed-off-by: tellang <tellang@daum.net>

SerenityZenDev commented on user-service/src/main/java/com/ticketing/userservice/application/service/UserService.java:
```diff
@@ -0,0 +1,83 @@
+	public Delete.Result softDeleteUser(Delete.Soft deleteDto) {
+		User user = repositoryHelper.findOrThrowNotFound(deleteDto.id());
+		user.softDelete(deleteDto.deleterId());
+		return deleteDtoFrom(userRepository.save(user));
```
@Transactional 내부에서 softDelete를 통해 상태변경을 진행 중인것으로 보이는데 save를 추가로 진행하는 이유가 있을까요?

tellang replied:
주신 의견 처럼 해당 부분은 진행 흐름상 불필요한 호출이 불필요한 부분인 것 같습니다!

SerenityZenDev replied:
해당 내용 이외에 별도로 크게 보이는 내용은 없어서 approve 하겠습니다 해당 내용 반영 이후에 merge 해주세요!

SerenityZenDev reviewed (commented):
유저 기능에 대해서 매우 잘 작성된 것으로 보이며 테스트도 적당해보입니다.
다만 한번의 PR에 너무 많은 변경사항이 담겨 있는 것으로 보여 PR 단위를 조금 더 작게 가져가는게 좋을 것 같습니다

SerenityZenDev reviewed (approved):
LGTM!

tellang committed:
:fire: softDeleteUser 메서드에서 불필요한 save 호출 제거

- User 엔티티는 영속성 컨텍스트에 의해 관리되므로, 명시적인 save 호출이 불필요함.
  - 트랜잭션 종료 시 엔티티 상태 변경이 자동으로 처리됨.

Signed-off-by: tellang <tellang@daum.net>
